### PR TITLE
Dynamic fragments

### PR DESCRIPTION
**Pull request by tptee**
`Fragment`s no longer rely on static routes for matching. This is the first and most important step in making consumer components code-splittable.

A few breaking changes fell out of this work:
- Route matching in `Fragment` now depends on the order of the routes of its sibling `Fragments`, not the order of the static routes. Check out the test diffs for the behavioral differences. The biggest thing to remember is to put index routes (a nested `/`) last in the `Fragment` hierarchy.
- Killed `RouterProvider` and `provideRouter` with 🔥 . We don't need them for matching hacks anymore, it's annoying boilerplate, and it [ _rerenders the entire app on location change_](https://github.com/FormidableLabs/redux-little-router/issues/182) (wtf was I thinking when I wrote that???). We just `connect()` `Fragment` and `Link` now.
- `connect()`ing our components resolves rendering bugs related to accessing mutating `context`. Redux and context don't play nicely [AT ALL](https://medium.com/@mweststrate/how-to-safely-use-react-context-b7e343eff076). We still use context to pass down `parentRoute` and `parentId`, but since there's rarely a need to dynamically change which route a `Fragment` matches on, we can safely assume that these values won't change after the first render (which is the prerequisite for using context safely with libraries like `react-redux`. We throw an error if `forRoute` changes for now. Supporting dynamic `forRoute` props would require a subscription system like that in `react-redux`, and ain't nobody got time for that.

### CODE REVIEW

**Comment by justrossthings** [comment]
🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥

File: src/components/fragment.js [hunk 1]
```diff
@@ -17,111 +73,78 @@ type Props = {
...
-        parentRoute: parentRoute &&
-          parentRoute !== '/' &&
-          parentRoute !== forRoute
-          ? `${parentRoute}${forRoute || ''}`
-          : forRoute,
-        parentId: this.id
-      };
-    }
+  constructor(props: Props, context: typeof Fragment.contextTypes) {
+    super(props, context);
```
Comment (kylecesmat) [inline src/components/fragment.js]: I don't believe you need `context` to be declared in `super()` if it's not being used in the constructor

File: src/components/fragment.js [hunk 2]
```diff
@@ -17,111 +73,78 @@ type Props = {
...
+        matchCache.add(parentId, currentRoute);
+      }
+    }
 
-export default relativePaths(Fragment);
+    return Children.only(children);
+  }
+}
+
+export default compose(
```
Comment (kylecesmat) [inline src/components/fragment.js]: this is some next-level composition 🔥

**Comment by kylecesmat** [comment]
👋  RouterProvider 🔥 🔥

File: src/components/fragment.js (hunk 1 above)
Comment (aweary) [inline src/components/fragment.js]: You also don't need to pass `props`, unless you're accessing `this.props` in the constructor. React will populate them both on the instance for you.

**Comment by faceyspacey** [comment]
@tptee you were thinking React Router does it. No harm no foul. Redux is king. Now we know. ...although there are a few things in `react-redux` that really should be revisited for even better perf. We'll see how Fiber affects `react-redux`.

**Pull request merged**